pkg/chunk/disk_cache_state.go

```diff
@@ -0,0 +1,297 @@
+		dc.cache.event(eventToUnstable)
+	}
+}
+
+type unstableDC struct {
+	baseDC
+	startTime time.Time
+	ioErrCnt  uint32
+	ioCnt     uint32
+
+	concurrency atomic.Int64
+}
+
+func (dc *unstableDC) state() int { return dcUnstable }
+
+func (dc *unstableDC) init(cs *cacheStore) {
+	dc.baseDC.init(cs)
+	dc.startTime = time.Now()
+}
+
+func (dc *unstableDC) onIOErr() {
+	atomic.AddUint32(&dc.ioCnt, 1)
+	atomic.AddUint32(&dc.ioErrCnt, 1)
+}
+
+func (dc *unstableDC) onIOSucc() {
+	atomic.AddUint32(&dc.ioCnt, 1)
+}
+
+func probeCacheKey(id, size int) string {
+	return fmt.Sprintf("%s/%02X/%v/%v_%v_%v", probeDir, id%256, id/1000/1000, id, 0, size)
+}
+
+func (dc *unstableDC) tick() {
+	go dc.probe()
+	go func() {
+		ticker := time.NewTicker(tickDurForUnstable)
+		defer ticker.Stop()
+
+		for {
+			select {
+			case <-dc.stopCh:
+				return
+			case <-ticker.C:
+				errCnt, ioCnt := atomic.LoadUint32(&dc.ioErrCnt), atomic.LoadUint32(&dc.ioCnt)
+				if ioCnt >= minIOSuccToNormal && float64(errCnt)/float64(ioCnt) <= maxIOErrPercentageToNormal {
+					dc.cache.event(eventToNormal)
+				} else if time.Since(dc.startTime) >= maxDurToDown {
+					dc.cache.event(eventToDown)
+				} else {
+					atomic.StoreUint32(&dc.ioErrCnt, 0)
+					atomic.StoreUint32(&dc.ioCnt, 0)
+				}
+			}
+		}
+	}()
+}
+
+func (dc *unstableDC) probe() {
+	page := NewPage(probeData)
+	defer page.Release()
+	cnt := 0
+
+	for {
+		select {
+		case <-dc.stopCh:
+			return
+		default:
+			cnt++
+			start := time.Now()
+			dc.doProbe(probeCacheKey(cnt, len(probeData)), page)
+			diff := probeDur - time.Since(start)
+			if diff > 0 {
+				time.Sleep(diff)
+			}
+		}
+	}
+}
+
+func (dc *unstableDC) doProbe(key string, page *Page) {
+	dc.cache.cache(key, page, true)
+	reader, err := dc.cache.load(key)
+	if err != nil {
+		return
+	}
+	defer reader.Close()
+	_, _ = reader.ReadAt(probeBuff, 0)
+	dc.cache.remove(key)
+}
+
+func (dc *unstableDC) beforeCacheOp() { dc.concurrency.Add(1) }
+func (dc *unstableDC) afterCacheOp()  { dc.concurrency.Add(-1) }
+
+func (dc *unstableDC) checkCacheOp() error {
+	if dc.concurrency.Load() >= maxConcurrencyForUnstable {
+		return errUnstableCoLimit
+	}
+	return nil
+}
+
+type downDC struct {
+	baseDC
+}
+
+func (dc *downDC) state() int          { return dcDown }
+func (dc *downDC) checkCacheOp() error { return errCacheDown }
+
+func (cache *cacheStore) event(eventType int) {
+	cache.stateLock.Lock()
+	defer cache.stateLock.Unlock()
+	state := cache.state.state()
+	switch state {
+	case dcNormal:
+		if eventType == eventToUnstable {
+			cache.state.stop()
+			cache.state = newDCState(dcUnstable, cache)
+		}
+	case dcUnstable:
+		switch eventType {
+		case eventToNormal:
+			cache.state.stop()
+			cache.state = newDCState(dcNormal, cache)
+		case eventToDown:
+			cache.state.stop()
+			cache.state = newDCState(dcDown, cache)
+		}
+	}
+	logger.Warnf("disk cache state change from %s to %s", diskStateNames[state], diskStateNames[cache.state.state()])
```

Comment:
add path into logging

Reply:
info

Reply:
done

Reply:
@jiefenghuang logger level?

Reply:
we had error msg when removeStore in cacheManager, so i think warning is enough

Reply:
@SandyXSD changed to info